{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <p style=\"text-align: center;\"> Introduction to Genetic Algorithms</p>\n",
    "#### <p style=\"text-align: center;\"> 40.319 Statistical and Machine Learning (SML)</p>\n",
    "#### <p style=\"font-size: 16px; text-align: center;\"> Authors: Md Ibrahim (1005972), Chua Jin Chou (1005988)</p>\n",
    "#### <p style=\"font-size: 16px; text-align: center;\"> 18 April 2024 | Year 3, Term 6 </p>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <p style=\"text-align: center;\"> Section 1: Motivation</p>\n",
    "\n",
    "As a kid, have you ever tried to solve a maze on pen and paper before? Mazes are commonly found inside children's activity books as a brain teaser to develop problem-solving skills. One will find themselves tracing lightly along a path with a pencil, before hastily erasing it off when the path turns out to be a dead end. As the mazes increase in difficulty, we will require multiple attempts before we can find the solution, often relying on the faded markings from our previous attempts to guide us to the correct path. As we learn from our past attempts and avoid making the same mistakes, we will eventually arrive at the correct path. \n",
    "\n",
    "In this report, we aim to showcase how genetic algorithms can help us solve a maze, in a similar method described above. You will also learn about the theory behind genetic algorithms and reinforcement learning and how genetic algorithms help us solve complex issues. Finally, you will see an application where we will be using a genetic algorithm that we have designed to solve a maze. You can also interact with the maze by setting the parameters of the maze solver, such as the size of the maze or the number of parents and observe how the algorithm solves the maze.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <p style=\"text-align: center;\"> Section 2: Context</p>\n",
    "In the latter half of SML, we learnt about several reinforcement learning methods, such as Deep Q learning and Markov Decision Process, which piqued our interest in self-learning algorithms. This led us to discover Genetic Algorithms (GA). While GA is not exactly a type of self-learning algorithm, it is considered to be an unspervised learning algorithm. We also felt that the topic was very interesting and its fundamentals could still be applicable in the realm of machine learning.\n",
    "\n",
    "GAs are considered to be a type of metaheuristic optimization algorithms. They belong to the larger class of Evolutionary Algorithms. GAs work by repeatedly assessing and developing new solutions from the best of past solutions to arrive at the most optimal one. This mimics how natural selection in nature works, where animals with desirable traits will have an advantage in the environment and get to pass on their genetics to future generations. \n",
    "\n",
    "A notable difference between reinforcement learning algorithms like Deep Q learning are aimed to achieve the maximum score, GAs are used to find the optimal or close-to-optimal solutions to complex problems, hence their use cases are strikingly different. \n",
    "\n",
    "In this Python notebook, we will be diving into the background of GAs: the logic behind GAs and why they work, how GA algorithm finds the optimal solutions, real world applications where they are used as well as a code demonstration that showcases how a simple GA algorithm will find an optimal solution through solving a maze solver example. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <p style=\"text-align: center;\"> Section 3: Fundamentals of Genetic Algorithms </p>\n",
    "\n",
    "This section will be dedicated to explaining the theory behind Genetic Algorithms.\n",
    "\n",
    "<div>\n",
    "<center><img src=\"Resources/local minima diagram.png\" width=\"60%\" title= \"Fig 1: Visualization of local and global optima on a graph\"/></center>\n",
    "<p style=\"text-align: center; ffont-size: 8 px; color:#9B9B9B;\"> Fig 1: Visualization of local and global optima on a graph </p>\n",
    "</div>\n",
    "\n",
    "This diagram above describes mathematical optimisation problems. In Optimisation problems, the goal is to find some optima (Glocal Maximum/Minimum) of the objective function within the search space or solution space. \n",
    "\n",
    "The objective function is modelled after your problem; it could be a formula to calculate profit based on multiple product sales, or something else. The objective function would exist within the solution space; in the diagram above, it is a 2D-space. For a 2D space, there would only be 1 parameter (the x-axis), so for instance it could be profit against number of phones sold. For many different variables or parameters, there would a n-D space (n-dimensional search space) where n is the number of variables you have. This could be the profit against the number of phone A sold and phone B sold and so on. This modelling of a real-world problem as a mathematical optimisation problem is widely used today.\n",
    "\n",
    "A solution to the problem is any point on the objective function, in the diagram, it is any point on the line (this is known as a feasible solution). There could be good solutions and better solutions. Not all solutions are created equal and while all solutions may produce satisfactory results, some solutions are simply better.\n",
    "\n",
    "In general, all optimisation algorithms have to conquer the usual challenges of optimisation. In the diagram above, note the “local search”, “local minimum” and “cycling” annotations. Those are the challenges of finding the optimal solution - how does one ensure that their algorithm is able to explore past these local optima and actually find the global optima? Think of it like getting stuck on a hill when you want to reach the peak of the mountain - we are more interested in the absolute best solution, and less interested in the other possible but satisfactory solutions. The global optima appears to be very simple to find in this graphical visualization but an optimisation algorithm faces challenges like getting stuck at a less ideal local optima solution.\n",
    "\n",
    "Even though the solution space and objective functions are laid out, it is difficult to actually find the global optima. The diagram above is a very simple example, but in practice, problems can have a search space higher than 3 dimensions such as in the 4D, 5D or even higher. It is very difficult to visually inspect the curves as we lack the capability for it. Thus, we require the use of specific algorithms to find these global optima for us. One such algorithm that does this is the topic of our project, Genetic Algorithms.\n",
    "\n",
    "<div>\n",
    "<center><img src=\"Resources/Overview of GA.png\" width=\"60%\" title= \"Fig 2: GA fundmanetals visualized\" /></center>\n",
    "<p style=\"text-align: center; ffont-size: 8 px; color:#9B9B9B;\"> Fig 2: Overview of genetics in Genetic Algorithms </p>\n",
    "</div>\n",
    "\n",
    "<div>\n",
    "<center><img src=\"Resources/GA overview diagram.png\" width=\"60%\" title= \"Fig 3: Overview of process of finding the optima solution using GA\" /></center>\n",
    "<p style=\"text-align: center; ffont-size: 8 px; color:#9B9B9B;\"> Fig 3: Overview of process of finding the optima solution using GAs </p>\n",
    "</div>\n",
    "\n",
    "Typically, GAs will start off by initialising a starting population, usually comprised of randomised genes or chromosomes. Afterwards, the initial evaluation function assesses the fitness level of the current population. Then, some **selection** criteria will select some members of the initial solution population as candidates to “mate” and produce “child” solutions. This selection operation usually allows only similar genes to be used and create children, and dissimilar solutions simply won’t be used. The selection criteria is also known as a **fitness function**. The child’s genes or chromosome would be decided by the **Crossover** operations which creates the gene for the child based on its parent’s genes. Children solutions will then undergo **Mutation** which makes them uniquely different and possibly better or worse than their parents.\n",
    "\n",
    "As a general note, these are merely notions of GAs. GAs operate probabilistically, hence some of these operations might not be required for the specific problem - as long as an algorithm contains the general flow of a GA, then it can be considered a GA still. Additionally, the high-level themes of a GA involve the notions of **\"Quality\"** and **\"Diversity\"**. Quality refers to how well, measured by the fitness function, a solution performs (how close to optimality it is) and Diversity refers to the different kinds of solutions there are. In general, they are trade-offs for each other - prioritising Quality will lead to a decrease in Diversity (usually, local optima is a hurdle here - it is \"too\" greedy) whereas vice vera, Diversity causes solutions to be too explorative and not actually get close to optimality. These are just high-level considerations. In practice, the balance between these 2 notions will depend on the specific implementation.\n",
    "\n",
    "### Section 3.1: Mating\n",
    "Mating, while not directly a step in the genetic algorithm, is the crux of the algorithm itself. It has 3 distinct steps below:\n",
    "\n",
    "Step 1: <span style=\"color:#8DB0C6;\">**Selection**</span>\n",
    "\n",
    "Generally, the selection operation finds 2 “suitable” parents that are able to produce a “valid” child. In practice, this is important as some genes might not be valid solutions for an optimisation problem, and hence cannot exist. Things like similarity between the parents or some other criteria are used to ensure the child is likely to be valid. However, if the optimisation problem is well-defined and has accounted for this, then the selection operation can simply allow all solutions to mate.\n",
    "\n",
    "- Elitism:  Variants of GAs try to replicate Natural Selection better, and it is done is the way shown. By adding another layer to the GA process, it could possibly result in better solutions or performance. However, adding more operations to the pipeline might require more computational power.\n",
    "- Elitism is a step that immediately kills off solutions that do not meet some fitness level. Adding this to the GA will force Quality over Diversity.\n",
    "\n",
    "Step 2: <span style=\"color:#8DB0C6;\">**Crossover**</span>\n",
    "\n",
    "Crossover operation determine which genes to inherit from parents. It is usually defined as cut-off points within the genes. In the example above, we showed ‘single point crossover’, meaning that the child would inherit one segment from each parent. In other applications, a ‘2 point cutoff’ can be applied as well, where the child inherits 2 disjoint parents from 1 parents and a middle component from the other parent.\n",
    "\n",
    "<div>\n",
    "<center><img src=\"Resources/Cross over cutoff.png\" width=\"60%\" title= \"Fig 4: Crossover cutoff diagram\" /></center>\n",
    "<p style=\"text-align: center; ffont-size: 8 px; color:#9B9B9B;\"> Fig 4: Crossover cutoff diagram </p>\n",
    "</div>\n",
    "\n",
    "Step 3: <span style=\"color:#8DB0C6;\">**Mutation**</span>\n",
    "\n",
    "<div>\n",
    "<center><img src=\"Resources/GA mutation.png\" width=\"60%\" title= \"Fig 5: Visualization of mutation in GA\" /></center>\n",
    "<p style=\"text-align: center; ffont-size: 8 px; color:#9B9B9B;\"> Fig 5: Visualization of mutation in GA </p>\n",
    "</div>\n",
    "\n",
    "Mutation simply adds some random changes to the child solution after crossover. This is important for the diversity aspect of GAs - *ensuring that each solution is exploring the problem in a slightly different way.* Mutation allows new solutions to differ slightly from the parent solutions, which increases the chance of finding the optima of the problem. Mutation can simply refer to any differntiation from the original parent solution, which can come in the form of either changing certain parts of the merged solution or expanding on the solution.\n",
    "\n",
    "### Section 3.2: Real World application of Genetic Algorithms\n",
    "\n",
    "Despite its downsides, GAs are very much relevant and widely used today.\n",
    "\n",
    "For instance, Professor Nuno Antunes Ribeiro, who researches at the Aviation Studies Institute in SUTD,  employs GAs in the scheduling problem of airplane landings. The traditional way to solve scheduling problems is via the famous Simplex method, which guarantees correctness, but is very slow. What GAs excel at are giving **good partial solutions**. The analogy for this is, would you finish a maze faster from the start of a maze or from somewhere closer to the end? Generally, it's much faster to finish the maze when you're much closer to the end. GAs do exactly this - you can stop a GA prematurely and use that solution regardless. In Professor Nuno's work, he has explained that GAs' flexibility aids other \"stronger\" algorithms like simplex (stronger in that they are provably correct) to complete the problem faster.\n",
    "\n",
    "Another real-life application of GAs is in architectural modelling. In the popular precision 3D modelling software Rhino, used by many architects worldwide, there exists a package called Galapagos Solver, which operates using a GA. This library leverages the GA's ability to give decent partial solutions, which are faster to achieve than waiting for another algorithm like simplex to complete.\n",
    "For instance, it is able to give partial solutions to problems with many, many constraints, perhaps too slow to even know what a valid solution looks like using simplex. In architecture, there are many constraints. For instance, a certain structure needs to be within a certain perimeter, with a certain density, with a certain surface area, and a certain number of contact points with the floor, and require certain amount of air flow through, etc. There could be many more requiremnts when designing a feasible architecutre structure. In situations like these, GAs excel at giving partial solutions that can already be used directly; while not the best possible solution, it is good enough for the architects to use and reference in their work.\n",
    "\n",
    "Essentially, a GA is used to find an initial feasible solution, and allow simplex to try and solve the problem from there, which could be repeated in a cyclic fashion, thus improving the overall performance when solving the problem.\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <p style=\"text-align: center;\"> Section 4: Genetic Algorithms Application - Maze solver </p>\n",
    "\n",
    "Now that you are equipped with the knowledge on the fundamentals of how GAs work, we will showcase our attempt at making a maze solver using the GA developed by our team. In this example, the chromosome that are being passed down to future generations are tuples containing x and y coordinates on the grid. \n",
    "\n",
    "In this demonstration, our algorithm will always start from the (n,n) square in the maze, where n is the length of the maze. The exit of the maze is at the (1,1) square. While each iteration of the maze will be unique, these two conditions will always hold true. \n",
    "\n",
    "When you run the code, you will see how there are different red-coloured paths exploring the maze. For simplicity sake, we called them <b>\"Worms\"</b>. These worms will spawn from the (n,n) square with each generation and you will notice how each worm gets longer in length. You will also see several scenarios where the algorithm will get stuck and how the worm will forcefully mutate in later generations to avoid the dead end.\n",
    "\n",
    "Our GA also utilizes some aspects of Deep Q Learning such as the cost function. This will be further elaborated on below.\n",
    "\n",
    "In the context of GAs, the objective function is the fitness function that determines the quality of the solution population. It could be seen as how well the solution “animal species” is “adapting” to its “environment”.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 4.1: Packages required\n",
    "We will be using pyamaze to generate a new maze each time, as well as the functions within for a visual representation of how the algorithm solves the maze."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyamaze import maze, agent, COLOR\n",
    "import random"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 4.2: Defining the Worm class\n",
    "This section will be detailing the Worm class. As mentioned before, the Worm class refers to an iteration of the path. Whenever a generate a new path, a new Worm object is initialized and it will take several inputs, namely:\n",
    "* <span style=\"font-size:16px; color:#8DB0C6;\"><b>id</b></span>: This attribute stores a unique numerical id for the path for easy identification\n",
    "* <span style=\"font-size:16px; color:#8DB0C6;\"><b>path</b></span>: This attritbute will contain a list that consists of all the coordinates in the path that the Worm object will follow. \n",
    "* <span style=\"font-size:16px; color:#8DB0C6;\"><b>pos</b></span>: This attribute will store the last known position of the worm\n",
    "* <span style=\"font-size:16px; color:#8DB0C6;\"><b>agent</b></span>: An agent refers to the visual entity we see traversing through the maze. It is an object unique to pyamaze and each Worm will have its own agent\n",
    "* <span style=\"font-size:16px; color:#8DB0C6;\"><b>score</b></span>: Each Worm will initialize with a score of maze height * maze width. More details on how the scoring system works will be given in later sections\n",
    "* <span style=\"font-size:16px; color:#8DB0C6;\"><b>DNA</b></span>: This attribute stores a dictionary that serves as the DNA of the Worm. The DNA will store information from past generations on which paths are more favourable. This DNA will directly influence the score of the Worm. More details on how the scoring system works will be given in later sections\n",
    "\n",
    "There are also several methods under the Worm class, namely:\n",
    "1. <span style=\"font-size:16px; color:#8DB0C6;\">**self.validPaths()**</span>: This function determines if a Worm can move from coordinate A to coordinate B.\n",
    "\n",
    "    It uitilizes the maze_map function, which stores all directions in the form a of a dictionary with keys \"N\", \"S\", \"E\", \"W\" and binary numbers 1 to indicate that the direction is valid and 0 to indicate the the direction is invalid. More details on how the dictionary works can be found in the documentation of the pyamaze package. \n",
    "\n",
    "    Invalid coordinates include:\n",
    "    1. previous coordinate in the path as the Worm is not allowed to move backwards in our algorithm\n",
    "    2. any coordinate blocked off by a wall\n",
    "\n",
    "    All invalid coordinates will be filtered and removed, as well as the previous coordinate in the path and return a list of valid coordinates that the Worm can move to from its current coordinate. Do note that when a dead end is encountered, this function will return an empty list \n",
    "\n",
    "    \n",
    "\n",
    "2. <span style=\"font-size:16px; color:#8DB0C6;\">**self.move()**</span>: This function is used to determine the next step the Worm will take. It determines all possible directions use **self.validPaths**. If there is more than one possible direction available for the Worm, this function will randomly select a direction and add it to the path. If **self.validPaths** returns [], this indicates that there are no valid paths and **self.move** will return \"Dead end\" instead. **self.path** and **self.pos** will then be updated accordingly.\n",
    "\n",
    "3. <span style=\"font-size:16px; color:#8DB0C6;\">**self.isValidPaths()**</span>: This function iterates through each coordinate in the list from **self.path** and recursively runs **self.validPaths** to check if the path is valid. If the path is valid, i.e. no walls blocking/backtracking, this function will return True else False\n",
    "\n",
    "4. <span style=\"font-size:16px; color:#8DB0C6;\">**self.mutate(steps)**</span>: This function will used to extend the path by x number of coordinates. The function recursively calls **self.move** for a set number of times, the default value is height + width of the maze. If **self.move** returns \"Dead end\", the mutation stops.\n",
    "\n",
    "   <span style=\"color:#FF5555\"> **This function serves as the \"mutation\" step of a Genetic Algorithm as it mutates the next generation by extending the length of the path.**</span>\n",
    "\n",
    "5. <span style=\"font-size:16px; color:#8DB0C6;\">**self.fitnessCheck()**</span>: This function determines the score of the path of the worm. The scoring system works like so:\n",
    "    1. Each worm initializes with a score of width + height\n",
    "    2. The lower the fitness score, the higher the probability of the worm leading being the solution.\n",
    "    3. Fitness of Worm = Sum of score of all coordinates in the path / length of path\n",
    "\n",
    "    **Rationale for this equation:**\n",
    "    We came up with this equation so that the priority of paths selected for crossbreeding will work like this: **Longer path that does not end at a dead end > shorter path that does not end at a dead end > dead end path**. \n",
    "    \n",
    "    Longer paths that do not lead to a dead end will naturally have a higher chance of containing part of the solution to the goal of the maze and should be priroitized for crossbreeding over any shorter paths. Dead end paths will have a significantly higher score than any path that does not lead to a dead end due to the **+1000 points** penalty for some coordinates, hence they will have a lower chance of being selected for crossbreeding. If all paths in a generation are dead ends, longer dead end paths will still be selected for crossbreeding and will be mutated on to find a potential solution. \n",
    "\n",
    "    We came up with this cost function after inspiration from how Deep Q learning works. We were hoping to encompass some self-learning attributes into the algorithm.\n",
    "    \n",
    "6. <span style=\"font-size:16px; color:#8DB0C6;\">**self.forcefullyMutate()**</span>: This function is used to forcefully rewrite the path of a Worm. This function is called to force future generation of Worms to explore other parts of the maze after they have been stuck in the same dead end. There are 4 steps in this function:\n",
    "\n",
    "    1. This function starts by checking if there are any alternative routes along the path to forcefully redirect the Worm to. \n",
    "    2. The function will then check if the path is \"valid\" using the **self.isValidPath**. \n",
    "    3. If the path is invalid (i.e. it passes through a wall or there is backtracking), the function will continuously pop the last coordinate in the path while checking if the path is valid after each pop. The function will stop removing coordinates when the path is finally valid \n",
    "    4. In the final step, the function will extend the path of the Worm by another height + width number of coordinates coordinates using **self.mutate**\n",
    "\n",
    "    <span style=\"color:#FF5555\">  **This function also serves as the \"mutation\" step of a Genetic Algorithm as it forces out a mutation of the path of future Worms.** </span>\n",
    "\n",
    "7. <span style=\"font-size:16px; color:#8DB0C6;\">**self.writeDNA()**</span>: This function updates the DNA dictionary of the worm. All paths leading up to a dead end, as well as the dead end itself will have their scores += 1000 pts, while paths that do not lead up to a dead end will keep a score of 1. \n",
    "\n",
    "8. <span style=\"font-size:16px; color:#8DB0C6;\">**self.combineDNA(p1, p2)**</span>: This function takes in the two parents of a Worm and combine their DNA dictionaries, taking the maximum value corresponding to each coordinate in their paths. The new Worm will then inherit this newly combined dictionary.\n",
    "\n",
    "9. <span style=\"font-size:16px; color:#8DB0C6;\">**self.crossbreedPath(parents)**</span>: This function will conduct crossover on a set of 2 parents to create a new path. It is called when a new Worm object is initialized to create a path for it. \n",
    "\n",
    "    The function takes in a list input called parents, which will include all the Worm objects chosen to be parents of the next generation after selection. A random set of 2 parents are selected to be the parents and their \"genetics\" (i.e. their coordinates in the path) are mixed around at random to produce a new path. To prevent mutation from occuring too early in the path, we introduce a variable c, which determines how many coordinates of the new path will be directly copied from the parents. \n",
    "\n",
    "    For example, 2 parent Worms both have a length of 10 coordinates. We let c be 70% of the length of the parent worms. Hence, the first 7 coordinates of the new Worm will be identical to the first 7 coordinates of one of the parents. Then, the subsequent 10-c = 10-7 = 3 coordinates will be randomly mixed and matched from the last 4 coordinates of the parents. \n",
    "\n",
    "    <div>\n",
    "    <center><img src=\"Resources/Crossover diagram.png\" width=\"65%\" alt = \"Fig 6: Crossover of Worms in GA\"/></center>\n",
    "    <p style=\"text-align: center; ffont-size: 8 px; color:#9B9B9B;\"> Fig 6: Crossover of Worms in GA </p>\n",
    "    </div>\n",
    "\n",
    "    The function will then check if the best fitness score from the most recent generation has improved compared to the best fitness score from 3 generations ago. If the previous fitness score is not strictly better than its predecessor from 3 generations ago, this would mean that bad \"genetics\" are being passed down. In other words, this means that it is likely the past 3 generations of worms have been stuck in a dead end loop and continuously passing down coordinates that lead to a dead end. Hence, the Worm will be forcefully mutated using **self.forcefullyMutate** to explore for new paths. This section ensures that the Worms will not continuously pass down dead end paths, as it is inevitable that after some generations, all future Worms will onto the same paths and end up in a dead end loop. \n",
    "\n",
    "    The function will then continuously check if the path is valid while mixing and matching coordinates and it will use a Boolean value to keep track of the validity of the path. Once the path is no longer valid, the last coordinate added to path will be removed (as it is invalid) and **self.mutation** is called to further extend the path.\n",
    "\n",
    "    <span style=\"color:#FF5555\">  **This function serves as the \"crossover\" step of this Genetic Algorithm as it merges part of the best parents from the previous generation to create the next generation. It also serves as the \"mutation\" step of this Genetic Algorithm as the path is mutated to differ from the parents using **self.mutate**.** </span>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Worm:\n",
    "    def __init__(self, idx, agent, path=[], DNA = {}):\n",
    "        self.id = idx\n",
    "        self.path = path\n",
    "        self.pos = agent.position if path == [] else path[-1]\n",
    "        self.agent = agent\n",
    "        self.score = h*w\n",
    "        self.DNA = DNA\n",
    "\n",
    "    def validPaths(self, pos):\n",
    "            #this function is used to determine all valid paths from a the worm's current position.\n",
    "            path_dict = m.maze_map[pos]\n",
    "            pm = []\n",
    "            x, y = pos\n",
    "            if path_dict[\"S\"] == 1 and x < h:\n",
    "                pm.append((x+1, y))\n",
    "            if path_dict[\"N\"] == 1 and x > 0:\n",
    "                pm.append((x-1, y))\n",
    "            if path_dict[\"W\"] == 1 and y > 0:\n",
    "                pm.append((x, y-1))\n",
    "            if path_dict[\"E\"] == 1 and y < w:\n",
    "                pm.append((x, y+1))\n",
    "            if len(pm) == 1 and pm[0] != goal:\n",
    "                deadends.append(pm[0])\n",
    "            return pm\n",
    "\n",
    "    def move(self):\n",
    "        #this function is used to determine which is the next step that the Worm will take in self.path.\n",
    "        pm = self.validPaths(self.path[-1]) #this will store the possible movements\n",
    "        if len(self.path) > 1:\n",
    "            check = self.path[-2]\n",
    "            pm.remove(check)\n",
    "        if pm == []: #this means that there are no valid positions available to move to, hence it is a dead end and the move function will stop here.\n",
    "            return \"Dead end\" \n",
    "        movement = pm[random.randint(0,len(pm)-1)]\n",
    "        \n",
    "        self.path.append(movement)\n",
    "        self.pos = self.path[-1]\n",
    "        return\n",
    "\n",
    "    \n",
    "    def isValidPath(self, path):\n",
    "        #this function determines if the sequence of coordinates in self.path is valid and will return True if so, else return False\n",
    "        if path == []: return True\n",
    "        for i in range(0,len(path)-1):\n",
    "            if path[i+1] not in self.validPaths(path[i]):\n",
    "                return False\n",
    "        return True\n",
    "    \n",
    "    def mutate(self, steps):\n",
    "        #this is the mutation function that will call the move function for steps to create a valid path for the worm\n",
    "        for i in range(steps):\n",
    "            t = self.move()\n",
    "            if t == \"Dead end\":\n",
    "                break\n",
    "        return\n",
    "    \n",
    "    def fitnessCheck(self):\n",
    "        #this function gives the fitness score of a path to the model path\n",
    "        i = 0\n",
    "        for j in range(1, len(self.path)): #iterates through each coordinate in the path\n",
    "            node = self.path[j]\n",
    "            self.score += self.DNA[node] #checks for the score in the DNA of the worm and add it to the score\n",
    "        self.score /= len(self.path) \n",
    "        return\n",
    "    \n",
    "    def forcefullyMutate(self):\n",
    "        #this function is called to forcefully mutate a path by modifying it's existing path and calling mutate from there. \n",
    "        p = self.path\n",
    "        #this for loop checks if there is an alternative path that can be branched out from the Worms original path\n",
    "        for i in range(0,len(self.path)-1):\n",
    "            t = self.validPaths(self.path[i])\n",
    "            t1 = p[i]\n",
    "            if i > 0:\n",
    "                t1 = p[i-1]\n",
    "            t2 = p[i+1]\n",
    "\n",
    "            if t1 in t:\n",
    "                t.remove(t1)\n",
    "            if t2 in t:\n",
    "                t.remove(t2)\n",
    "            if len(t) >0:\n",
    "                self.path = self.path[:i]\n",
    "                self.path.append(t[0])\n",
    "                self.pos = self.path[-1]\n",
    "                break\n",
    "            self.pos = self.path[-1]\n",
    "        \n",
    "        #this is a sanity check to ensure that the Worm's path is still valid\n",
    "        valid = self.isValidPath(self.path)\n",
    "        \n",
    "        #this for loops will recursively remove the last coordinate in a Worm's path list until the path is valid\n",
    "        while not valid:\n",
    "            self.path.pop()\n",
    "            valid = self.isValidPath(self.path)\n",
    "            \n",
    "        self.mutate(h+w)\n",
    "        return\n",
    "    \n",
    "    def writeDNA(self):\n",
    "        dead_path = False\n",
    "        for i in range(-1, -len(self.path)-1, -1):\n",
    "            t = self.validPaths(self.path[i])\n",
    "            if len(t) == 1: #this means dead end\n",
    "                self.DNA[self.path[i]]=self.DNA.get(self.path[i],0)+1000\n",
    "                dead_path = True\n",
    "            elif len(t) == 2:\n",
    "                if dead_path:\n",
    "                    self.DNA[self.path[i]]=self.DNA.get(self.path[i],0)+1000\n",
    "                else:\n",
    "                    self.DNA[self.path[i]]=self.DNA.get(self.path[i],1)\n",
    "            elif len(t) >2:\n",
    "                for coor in t:\n",
    "                    if coor not in self.DNA.keys() or self.DNA[coor] < 1000:\n",
    "                        dead_path = False\n",
    "                        self.DNA[self.path[i]]=self.DNA.get(self.path[i],1)\n",
    "                        break\n",
    "                if dead_path:\n",
    "                    self.DNA[self.path[i]]=self.DNA.get(self.path[i],0)+1000\n",
    "        return\n",
    "\n",
    "    def combineDNA(self, p1, p2):\n",
    "        newDNA = p1.DNA\n",
    "        for i in p2.DNA.keys():\n",
    "            newDNA[i] = max(newDNA.get(i, 0), p2.DNA[i])\n",
    "        return newDNA\n",
    "\n",
    "    def crossbreedPath(self, parents):\n",
    "        p1, p2 = random.sample(parents, 2) #initializing parent 1 and parent 2\n",
    "\n",
    "        #finding the shortest path\n",
    "        shortest = min(len(p1.path), len(p2.path))\n",
    "\n",
    "        #this variable decides how early on in the path does mutation occur.\n",
    "        c = max(1,shortest//2 + shortest//4) #Change the logic here if it mutates too early in the path\n",
    "\n",
    "        #the first c coordinates from either p1 or p2 will be directly copied and added to the path of this worm object\n",
    "        #mutation will then occur after the first c coordinates.\n",
    "        self.path = random.sample([p1.path[:c],p2.path[:c]],1)[0]\n",
    "        \n",
    "        #this step combines the DNA of p1 and p2\n",
    "        self.DNA = self.combineDNA(p1,p2)\n",
    "\n",
    "        #this section compares best score from most recent generation to that of 3 generations ago and determines if a forceful nutation is required.\n",
    "        if len(best) > 3:\n",
    "            now = best[-1]\n",
    "            past = best[-4]\n",
    "            if now >= past:\n",
    "                self.forcefullyMutate()\n",
    "            \n",
    "        #this section is the crossover step between the paths of the two parent Worms.\n",
    "        valid = True\n",
    "        while valid and c < shortest:\n",
    "            x = random.sample([p1.path[c],p2.path[c]],1)\n",
    "            self.path.append(x[0])\n",
    "            self.pos = self.path[-1]\n",
    "\n",
    "            c += 1\n",
    "            valid = self.isValidPath(self.path)\n",
    "            if not valid:\n",
    "                self.path.pop()\n",
    "                self.pos = self.path[-1]\n",
    "        to_add = shortest - c\n",
    "        \n",
    "        #this section covers the mutation step, which occurs right after crossover of parent's genetics.\n",
    "        self.mutate(to_add+h+w)\n",
    "        return\n",
    "\n",
    "### END OF WORM CLASS ###"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 4.3: Other important functions\n",
    "This section will be dedicated to detailing other important functions. \n",
    "\n",
    "1. <span style=\"font-size:16px; color:#8DB0C6;\"><b>startFresh()</b></span>: This function is used at the first initialization of the Worms. It is used to create a list called path that includes gen times of [(h,w)], where gen is the number of worms per generation and (h,w) are the starting coordinates of the maze\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def startFresh():\n",
    "    path = []\n",
    "    for i in range(gen):\n",
    "        path.append([(h,w)])\n",
    "    return path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. <span style=\"font-size:16px; color:#8DB0C6;\"><b>selection(worms)</b></span>: This function is used to select the best fit worms from the input generation of worms. There is a preset pool size for number of parents determined in main(). The function starts by initializing two lists: **elites** will store the Worms with the best fitness score, while **scores** will store the scores of each Worm. \n",
    "\n",
    "    As an example, we let the maximum number of possible parents be 4. The first 4 Worms will be automatically added into the **elites** list and they will be constantly replaced by other Worms with a better fitness score. Once all the worms in the input generation have been assessed and compared with each other, the remaining Worms in the **elites** list will be the Worms with the best fitness score. The function will then return **elites**. \n",
    "\n",
    "    <span style=\"color:#FF5555\"> **This type of selection incorporates elitism as it drives out lower quality solutions.**</span> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def selection(worms):\n",
    "    elites = []\n",
    "    scores = []\n",
    "\n",
    "    for w in worms:\n",
    "        w.fitnessCheck()\n",
    "        if len(elites) < parents: #parents here is a global variable found in main(). It determines the poolsize of parent Worms.\n",
    "            elites.append(w)\n",
    "            scores.append(w.score)\n",
    "        else:\n",
    "            worst = max(scores)\n",
    "            if worst > w.score:\n",
    "                s = scores.index(worst)\n",
    "                scores[s] = w.score\n",
    "                elites[s] = w\n",
    "    return elites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. <span style=\"font-size:16px; color:#8DB0C6;\"><b>createTracePath(worms)</b></span>: This function is used to convert the path list in **self.path** to a dictionary path for the agent attribute in Worm. A dictionary is used as the path input for the agent to follow and it is path that we see on screen as the Worms traverse the maze. More details on the formatting for the path dictionary for a maze agent can be found in the pyamaze documentation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def createTracePath(worms):\n",
    "    trace = {}\n",
    "    for w in worms:\n",
    "        trace[w.agent] = w.path\n",
    "    return trace"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. <span style=\"font-size:16px; color:#8DB0C6;\"><b>initializeWorms(x)</b></span>: This function is called once at the start of the algorith. It is used to initialize the first generation of worms. The number of worms per generation is dictated by the global variable **gen** in main(). This function creates **gen** number of Worm objects, calling **self.mutate** and **self.writeDNA** to create the first paths and first DNA dictionaries respectively. A variable x is used to keep track of the unique numerical ID of each Worm, and it increments by 1 at the end of each iteration in the for loop. After the first generation of Worms have been initialized, we create a list **trace** which will contain the path dictionaries for each agent. We then use **m.tracePath** to run the visualization of the simulation. Note that **m** is the variable storing the maze object when it is generated.\n",
    "\n",
    "    Finally, the path of each Worm is updated with a [(h,w)], to indicate the starting coordinate of the Worm. The function then returns the unique numerical id, **x**, and the list of Worm objects, **worm**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def initializeWorms(x):\n",
    "    #x stores the numerical id of the next worm\n",
    "    worms = []\n",
    "    for i in range(gen):\n",
    "        t = Worm(idx=x, agent = agent(m, footprints = True, color = COLOR.red, filled = True), path = [(h,w)])\n",
    "        t.mutate(h+w)\n",
    "        t.writeDNA()\n",
    "        worms.append(t)\n",
    "        x+= 1\n",
    "        t = None\n",
    "    trace = createTracePath(worms)\n",
    "    m.tracePath(trace, kill = True, delay = 50)\n",
    "\n",
    "    for i in worms:\n",
    "        i.path = [(h,w)] + i.path\n",
    "    return x, worms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5. <span style=\"font-size:16px; color:#8DB0C6;\"><b>traverseMaze(x, worms)</b></span>: This function is used for generating subsequent generations of Worms after the first generation. Calling the function once generates one generation worth of Worms. Hence, to generate 20 generations of worms, this function will be called 19 times as **initializeWorms** is called once at the start of the algorithm. It takes in the input **x** and **worms**, which stores the unique numerical id and a list of Worm objects respectively. \n",
    "\n",
    "    This function starts by filtering out the Worms from the previous generation and storing it in **parents** as a list. It then runs selection on this list to determine the best fit Worms out in the previous generation. <span style=\"color:#FF5555\"> Do note that this is the \"selection\" step in GA. </span> **found_paths** is a boolean value determining if a solution to the maze problem has been found, where True means a solution has been found and vice versa. \n",
    "\n",
    "    The function also finds the lowest score from the most recent generation and stores it in global variable **best**. This allows the function **self.crossbreedPaths** to compare fitness scores between past generations to determine if a forced mutation should occur.\n",
    "    \n",
    "    The function then generates another **gen** number of Worms that are the next generation. Each worm will have their paths generated by calling **self.crossbreedPath** on each Worm, which takes **parents** as the input list. The function will also check if (1,1) is inside the path of the Worm, as this would indicate that a solution has been found and the algorithm should stop here. If (1,1) was found in a Worm object, **found_path** will be set to True. <span style=\"color:#FF5555\"> Do note that this is the \"crossover\" and \"mutation\" steps in the GA, specifically running from the **self.crossbreedPath** function. </span> \n",
    "\n",
    "    In the last section, the function calls **createTracePath** to generate path dictionaries for each Worm and plays the animation using **m.tracePath**. The function will return the unique id **x**, the list containing all Worms **worms**, and the boolean value **found_path**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def traverseMaze(x, worms):\n",
    "    parents = worms[-1: -gen-1:-1]\n",
    "    chosen_ones = selection(parents)\n",
    "    found_path = False #determines if a solution has been found in this generation of Worms. \n",
    "\n",
    "    score = []\n",
    "    for i in chosen_ones:\n",
    "        score.append(i.score)\n",
    "    best.append(min(score)) \n",
    "\n",
    "    for i in range(gen):\n",
    "        u = Worm(idx=x, agent = agent(m, footprints = True, color = COLOR.red, filled = True), path = [(h,w)])\n",
    "        u.crossbreedPath(chosen_ones)\n",
    "        if u.path[0] != (h,w): #ensures that the final worm does actually start from coordinate (h,w). This is a fix for a bug. \n",
    "            u.path = [(h,w)] + u.path\n",
    "        if goal in u.path: #this line checks if (1,1) is in the worm's path, as this would indicate that the Worm's path is the solution to the maze.\n",
    "            found_path = True\n",
    "        u.writeDNA() #updates the DNA dictionary for future generations\n",
    "        worms.append(u)\n",
    "        u = None\n",
    "        x+= 1\n",
    "\n",
    "    newest_gen = worms[-1:-gen-1:-1]\n",
    "    trace = createTracePath(newest_gen) #creates path dictionaries for the agents \n",
    "    m.tracePath(trace, kill = True, delay = 50) #plays the animation\n",
    "    \n",
    "    return x, worms, found_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6. <span style=\"font-size:16px; color:#8DB0C6;\"><b>main()</b></span>: This is the main() function, where all the key variables are declared from. These variables are declared to be global as they are used frequently throughout different functions. Green coloured variables are variables that the user can interact with to observe difference between each run. The key variables are: \n",
    "\n",
    "    - <span style=\"font-size:16px; color:#8AE07A;\"><b>h</b></span>: This variable should be a positive integer denoting the height of the maze. The default value is 12 for demonstration purposes. \n",
    "    \n",
    "    - <span style=\"font-size:16px; color:#8AE07A;\"><b>w</b></span>: This variable should be a positive integer denoting the width of the maze. The default value is 12 for demonstration purposes. \n",
    "    - <span style=\"font-size:16px; color:#8AE07A;\"><b>gen</b></span>: This variable should be a positive integer which determines the number of Worms per generation. The default value is 10 for demonstration purposes. \n",
    "    - <span style=\"font-size:16px; color:#8AE07A;\"><b>parents</b></span>: This variable should be a positive integer which determines the maximum number of parents eligible for **crossover** from the **selection** stage. The default value is 2 for demonstration purposes. \n",
    "    - <span style=\"font-size:16px; color:#8AE07A;\"><b>max_generations</b></span>: This variable should be a positive integer which states the maximum number of generations before the algorithm terminates (i.e. give up looking for a solution). This is not to be confused with **gen**, which determines the number of Worms per generation. The default value is 30 for demonstration purposes. \n",
    "    - <span style=\"font-size:16px; color:#8DB0C6;\"><b>goal</b></span>: This variable is a tuple containing the x,y coordinates of the goal. This should be always set at (1,1) and should not be changed by the user.\n",
    "    - <span style=\"font-size:16px; color:#8DB0C6;\"><b>deadends</b></span>: This is an empty list storing coordinates of all deadends. Whenever a Worm reaches a dead end, the last coordinate will be added to this list. This is used to check if a specific coordinate is a dead end and can be considered as a \"blacklist\" of coordinates that Worms should never visit.\n",
    "    - <span style=\"font-size:16px; color:#8DB0C6;\"><b>best</b></span>: This is also an empty list that stores the best fitness score of each generation of Worms. This list is used to compare between past generations of Worms to determine if the Worms are in a dead end loop. More details were provided in previous sections. \n",
    "    - <span style=\"font-size:16px; color:#8DB0C6;\"><b>found_path</b></span>: This variable stores a boolean value that determines if a solution path has been found. When this variable becomes True, that means a solution has been found and the algorithm should stop running. By default, this variable is set to False.\n",
    "\n",
    "    The function starts by creating a maze object and storing it in **m**. A maze is then created using the **CreateMaze()** method. It then generates the first batch of Worms using **initializeWorms** and continues searching for a solution in the maze by iteratively calling **traverseMaze** for a maximum of **max_generations** or until the solution has been found. \n",
    "\n",
    "    Once the preceeding for loop ends, a maze agent is created with its colour set to blue. This maze agent will trace the solution to the maze, which can be found using **m.path**. **m.run()** runs the entire simulation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    global h, w, gen, goal, deadends, m, parents, best\n",
    "\n",
    "    ### FEEL FREE TO CHANGE THE VARIABLES HERE, INPUTS SHOULD ONLY BE POSITIVE INTEGERS###\n",
    "    h = 12 #height of maze\n",
    "    w = 12 #width of maze\n",
    "    gen = 10 #Worms per generation\n",
    "    parents = 2 #eligible number of parents per generation\n",
    "    max_generations = 30 # maximum number of generations before termination of algorithm\n",
    "\n",
    "    ### DO NOT CHANGE THESE VARIABLES###\n",
    "    goal = (1,1) #target coordinate of maze\n",
    "    deadends = []\n",
    "    best = []\n",
    "    found_path = False\n",
    "\n",
    "    #initialize maze\n",
    "    m=maze(h,w)\n",
    "    m.CreateMaze()\n",
    "\n",
    "    x = 0 #stores unique numerical id of Worms\n",
    "    x, worms = initializeWorms(x)\n",
    "\n",
    "    for i in range(max_generations): \n",
    "        if found_path != True:\n",
    "            x, worms, path = traverseMaze(x, worms)\n",
    "            found_path = path\n",
    "        else:\n",
    "            print(\"FOUND PATH\")\n",
    "            break\n",
    "\n",
    "    #final path \n",
    "    final = agent(m, color = COLOR.blue, footprints = True, filled = True)\n",
    "    m.tracePath({final:m.path}, delay = 100)\n",
    "    m.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 4.4: Running the maze\n",
    "\n",
    "This is where the magic happens. Feel free to change some of the variables mentioned in the main() function and compare the differences between each run. <span style=\"color:#FF5555\"> Remember to restart all kernels after each run of the algorithm. </span>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FOUND PATH\n"
     ]
    }
   ],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 4.5: Reflection and analysis of the algorithm's performance\n",
    "\n",
    "<span style=\"font-size:16px; color:#8DB0C6;\"> <u> **Review of the model's performance** </u></span>\n",
    "\n",
    "The GA is able to solve consistently for mazes of size 20 x 20, with a maximum number of 30 generations and 10 Worms per generation. After which, the algorithm could still solve for larger mazes, with the largest successful solution found for a 25 x 25 maze. The machine learning aspect in this algorithm comes from how only the best Worms are chosen to \"mate\" and create the next generation to traverse the maze.\n",
    "\n",
    "However, its performance drops drastically as the maze increases and it frequently gets stuck in dead end loops. The lag also builds up the longer the simulation runs due the increasing number of Worm objects. Ultimately, it is not a sure-win method and it serves better as a visual demonstration showcasing our understanding of GA and our novice attempt at utilizing it's concept in a real world application. \n",
    "\n",
    "You are encouraged to try out different combinations of variables in the GA. Do let us know if you were able to achieve a successful solve for a maze larger than 25 x 25!\n",
    "\n",
    "<span style=\"font-size:16px; color:#FF5555\"><u> **Challenges faced when developing fitness function** </u></span>\n",
    "\n",
    "We would also like to dedicate this section to reflect on the biggest challenge we faced while designing this GA, which was coming up with an original working fitness function. Our previous attempt was to compare the number of coordinates in a Worm object to the coordinates in the solution path derived from **m.path** and scale the fitness score from this. In other words, paths with more coordinates matching in the solution path will have a better score and hence a higher chance to be selected for crossover. \n",
    "\n",
    "While this fitness function was useful, we decided to scrap it as it is not realistic in the real world for us to know what is the ideal solution to a maze. If the ideal solution has already been found, what are we even trying to solve? Hence, after many trials and experiments, we finally settled on our current fitness function, which penalizes paths based on the number of coordinates leading up to a dead end and while rewarding longer paths. This may not be the best possible fitness function, however we are still very much satisfied with its performance.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 4.6: Future improvements \n",
    "\n",
    "Some possible future improvements includes:\n",
    "\n",
    "1. <span style=\"font-size:16px; color:#8DB0C6;\">**Improved blacklisting function**</span>: As of right now, the blacklisting only works for paths leading directly to a dead end. However, for coordinates with >= 2 alternative paths will never be blacklisted. There are many scenarios where all paths from these coordinates always lead to a dead end, however our algorithm does not account for that. Hence, this is a possible point for improvement.\n",
    "\n",
    "2. <span style=\"font-size:16px; color:#8DB0C6;\">**Removing old Worms**</span>: The longer the algorithm runs, the laggier the animation. This is due to the increasing number of agent object in the maze. There should only be a set number agents needed at one time and it should not be increasing with each generation. Future improvements could include removing the Worm objects and their maze agents after they have finish running. \n",
    "\n",
    "3. <span style=\"font-size:16px; color:#8DB0C6;\">**Optimizing of code**</span>: There are certain places in the code that are not fully optimized. We could reduct the number of functions and methods by combining them to make the code more compact and lean. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 5: References\n",
    "This section will be detailing the online sources we took references from.\n",
    "\n",
    "Source for Figure 1: [Exploring the Efficiency of Tabu Search in Optimization Problems](https://medium.com/the-modern-scientist/exploring-the-efficiency-of-tabu-search-in-optimization-problems-c72833873c31)\n",
    "\n",
    "Source for Figure 2,3,4,5,6: Made by our team\n",
    "\n",
    "Documentation for pyamaze package: https://github.com/MAN1986/pyamaze\n",
    "\n",
    "<u>**Sources used for understanding GAs**</u>\n",
    "\n",
    "Source 1 for GA explanation: [What is Genetic Algorithm?](https://www.mathworks.com/help/gads/what-is-the-genetic-algorithm.html)\n",
    "\n",
    "Source 2 for GA explanation: [Genetic Algorithms](https://www.geeksforgeeks.org/genetic-algorithms/)\n",
    "\n",
    "Source 3 for GA explanation: [What Are Genetic Algorithms? Working, Applications, and Examples](https://www.spiceworks.com/tech/artificial-intelligence/articles/what-are-genetic-algorithms/)\n",
    "\n",
    "Source 4 for GA explanation: [Introduction to Genetic Algorithm](https://medium.com/geekculture/introduction-to-genetic-algorithm-d417119040b7)\n",
    "\n",
    "Source 5 for GA explanation: [Genetic Algorithm](https://www.sciencedirect.com/topics/mathematics/genetic-algorithm)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
